core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java

```diff
@@ -1588,6 +1588,100 @@ private void doTestFindOffsetByTimestamp(long ts, long startOffset, int targetLe
+
+        List<EpochEntry> epochEntries = new ArrayList<>();
+        epochEntries.add(new EpochEntry(0, 0L));
+        epochEntries.add(new EpochEntry(1, 20L));
+        epochEntries.add(new EpochEntry(3, 50L));
+        epochEntries.add(new EpochEntry(4, 100L));
+        epochEntries.add(new EpochEntry(5, 200L));
+        checkpoint.write(epochEntries);
+        LeaderEpochFileCache cache = new LeaderEpochFileCache(tp, checkpoint, scheduler);
+        when(mockLog.leaderEpochCache()).thenReturn(Option.apply(cache));
+
+        long timestamp = time.milliseconds();
+        RemoteLogSegmentMetadata metadata0 = new RemoteLogSegmentMetadata(new RemoteLogSegmentId(tpId, Uuid.randomUuid()),
+                0, 99, timestamp, brokerId, timestamp, 1024, Optional.empty(), RemoteLogSegmentState.COPY_SEGMENT_FINISHED, truncateAndGetLeaderEpochs(epochEntries, 0L, 99L));
+        RemoteLogSegmentMetadata metadata1 = new RemoteLogSegmentMetadata(new RemoteLogSegmentId(tpId, Uuid.randomUuid()),
+                100, 199, timestamp + 1, brokerId, timestamp + 1, 1024, Optional.empty(), RemoteLogSegmentState.COPY_SEGMENT_FINISHED, truncateAndGetLeaderEpochs(epochEntries, 100L, 199L));
+        // Note that the metadata2 is in COPY_SEGMENT_STARTED state
+        RemoteLogSegmentMetadata metadata2 = new RemoteLogSegmentMetadata(new RemoteLogSegmentId(tpId, Uuid.randomUuid()),
+                100, 299, timestamp + 2, brokerId, timestamp + 2, 1024, Optional.empty(), RemoteLogSegmentState.COPY_SEGMENT_STARTED, truncateAndGetLeaderEpochs(epochEntries, 200L, 299L));
+
+        when(remoteLogMetadataManager.listRemoteLogSegments(eq(tpId), anyInt()))
+            .thenAnswer(ans -> {
+                int epoch = ans.getArgument(1);
+                if (epoch < 4) {
+                    return Collections.singletonList(metadata0).iterator();
+                } else if (epoch == 4) {
+                    return Arrays.asList(metadata1, metadata2).iterator();
+                } else {
+                    throw new IllegalArgumentException("Unexpected call!");
+                }
+            });
+        // Different (timestamp, offset) is chosen for remote and local read result to assert the behaviour
+        // 9999 -> refers to read from local, 999 -> refers to read from remote
+        FileRecords.TimestampAndOffset expectedLocalResult = new FileRecords.TimestampAndOffset(timestamp + 9999, 9999, Optional.of(Integer.MAX_VALUE));
+        FileRecords.TimestampAndOffset expectedRemoteResult = new FileRecords.TimestampAndOffset(timestamp + 999, 999, Optional.of(Integer.MAX_VALUE));
+        Partition mockFollowerPartition = mockPartition(tpId);
+
+        LogSegment logSegment = mock(LogSegment.class);
+        mockLogSegment(logSegment, 50L, timestamp, null);
+        LogSegment logSegment1 = mock(LogSegment.class);
+        mockLogSegment(logSegment1, 100L, timestamp + 1, expectedLocalResult);
```

Comment:
nit: We can let `mockLogSegment` return `LogSegment` object, and do the mock inside the method. ex:
```
private LogSegment mockLogSegment(long baseOffset,
                                long largestTimestamp,
                                FileRecords.TimestampAndOffset timestampAndOffset) throws IOException {
        LogSegment logSegment = mock(LogSegment.class);
        when(logSegment.baseOffset()).thenReturn(baseOffset);
        when(logSegment.largestTimestamp()).thenReturn(largestTimestamp);
        if (timestampAndOffset != null) {
            when(logSegment.findOffsetByTimestamp(anyLong(), anyLong()))
                    .thenReturn(Optional.of(timestampAndOffset));
        }
        return logSegment;
    }
```